Use TImeLimitedMatcher for matching regular expressions

airadier opened the pull request:
We were suffering severe high CPU usage issues from time to time in our 2 nodes Bitbucket server. After some investigation, we detected that threads matching regular expressions from yacc plugin were causing the high CPU usage. This lead us to the RegEx DoS https://www.owasp.org/index.php/Regular_expression_Denial_of_Service_-_ReDoS problem, which can be reproduced easily by using an evil regex and a specially crafter commit message.

In order to avoid DoS with badly crafted regular expressions, use TImeLimitedMatcher from https://www.exratione.com/2017/06/preventing-unbounded-regular-expression-operations-in-java/ to set matching timeout to 2 seconds (default). If the match fails to finish before the timeout, the commit will be rejected and an error message logged.

Commit:
Use TImeLimitedMatcher for matching regular expressions

In order to avoid DoS with badly crafted regular expressions, use TImeLimitedMatcher from https://www.exratione.com/2017/06/preventing-unbounded-regular-expression-operations-in-java/ to set matching timeout to 2 seconds (default). If the match fails to finish before the timeout, the commit will be rejected and an error message logged.

Commit:
Use openjdk8

Commit:
Fix Curl following redirects

Commit:
Fix tests

Commit:
Merge branch 'fix-oraclejdk8' into regex-match-timeout

Fix travis-ci build issues with oraclejdk8, and fix unittests

sford commented on src/main/java/com/isroot/stash/plugin/YaccHook.java:
```diff
@@ -66,7 +66,16 @@ private RepositoryHookResult handleRepositoryPush(
+        RepositoryHookResult result;
+        try {
+            result = yaccService.check(context, repositoryPushHookRequest, settings);
+        } catch (TimeLimitedMatcherFactory.RegExpTimeoutException e) {
```
Ideally we should have a unit test for this to ensure `TimeLimitedMatcherFactory` is working as intended and the exception is being caught with a message being shown to the committer.

[Similar to other question](https://github.com/sford/yet-another-commit-checker/pull/223#issuecomment-532205767)... is there an example regex and commit message that can trigger regex dos reliably for use for a unit test?

sford replied:
Currently the other commit regex tests are located at in [YaccServiceImplTest](https://github.com/sford/yet-another-commit-checker/blob/75c5cd0c65fafa1bbcc7b410408dbb44915c24fa/src/test/java/ut/com/isroot/stash/plugin/YaccServiceImplTest.java#L321-L346)

airadier replied:
Hi @sford , I'll take a look and include a unit test. In our case, we didn't find what regex(s) were causing the issues. We only found that there were many threads causing high CPU usage. Our first approach was adding some additional logs to measure the time taken to evaluate the regex, but as we were already using our custom plugin, we went directly for the fix.

In my tests, using:
`^(a+)+$ `
as pattern, and a commit message like:
`aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaX`

would take more than 2 seconds and throw the exception after making the code changes. Testing in https://regex101.com/ that regex with `aaaaaaaaaaaaaaaaaX` as input will report 393216 steps and 223 ms, and adding additional "a" to the input string will be reported as "catastrophic backtrack". Apparently every additional "a" in the input will double the number of steps.

airadier replied:
Added a unit test for this example regular expression. One thing to note: I was using JDK 12 in my latest builds instead of JDK 8 and I was not able to make the test fail! I just noticed that using JDK 8 the problem can be reproduced, but it has been fixed in JDK 12 (or maybe an earlier version). So that is good news! We should keep the test for now, but maybe the fix is not needed anymore when using a newer Java version, or at least for this specific regex (but others might also cause catastrophic backtrack, I didn't find anything relating to this issue being fixed in newer Java versions)

airadier replied:
According to https://stackoverflow.com/questions/53048859/is-java-redos-vulnerable, Java 11 is not vulnerable to this regex, but changing it to `((a+)+)+` raises the issue again. So I am using a more complex expression to make the test work on newer JDK versions.

sford commented on src/main/java/com/isroot/stash/plugin/TimeLimitedMatcherFactory.java:
```diff
@@ -0,0 +1,146 @@
+/**
+ * A factory class used to generate Matcher instances that will throw if in use after their
+ * timeout.
+ */
```
Can you add attribution to javadoc here? Maybe URL pointing to blog post code is from?

airadier replied:
Sure, I'll do it and update the PR.

airadier replied:
Done

Commit:
Fixes for PR

- Add UnitTest for catastrophic backlog
- Add source URL for TimeLimitedMatcherFactory

Commit:
Use more complex regex to make test fail in JDK 11+
